compiler/hie-core/src/Development/IDE/Core/Compile.hs

```diff
@@ -310,29 +311,61 @@ getModSummaryFromBuffer fp contents dflags parsed = do
     , ms_parsed_mod   = Nothing
     }
 
+-- | Run CPP on a file
+runCpp :: DynFlags -> FilePath -> Maybe SB.StringBuffer -> IO SB.StringBuffer
+runCpp dflags filename contents = withTempDir $ \dir -> do
+    let out = dir </> takeFileName filename <.> "out"
+    case contents of
+        Nothing -> do
+            -- Happy case, file is not modified, so run CPP on it in-place
+            -- which also makes things like relative #include files work
+            -- and means location information is correct
+            doCpp dflags True filename out
+            liftIO $ SB.hGetStringBuffer out
+
+        Just contents -> do
+            -- Sad path, we have to create a version of the path in a temp dir
+            -- __FILE__ macro is wrong, ignoring that for now (likely not a real issue)
+
+            -- Relative includes aren't going to work, so we fix that by adding to the include path.
+            let addSelf (IncludeSpecs quote global) = IncludeSpecs (takeDirectory filename : quote) global
+            dflags <- return dflags{includePaths = addSelf $ includePaths dflags}
+
+            -- Location information is wrong, so we fix that by patching it afterwards.
+            let inp = dir </> "___HIE_CORE_MAGIC___"
+            let f x = if SB.atEnd x then Nothing else Just $ SB.nextChar x
+            liftIO $ writeFileUTF8 inp (unfoldr f contents)
+            doCpp dflags True inp out
+
+            -- Fix up the filename in lines like:
+            -- # 1 "C:/Temp/extra-dir-914611385186/___HIE_CORE_MAGIC___"
+            let tweak x
+                    | Just x <- stripPrefix "# " x
+                    , "___HIE_CORE_MAGIC___" `isInfixOf` x
```

Comment:
Ugh, I feel like there must be a better way to handle this. This seems like a problem that every LSP server for languages that support CPP has to deal with.

Reply:
Agreed - but inclined to leave it as is for now. It works. If there's a better way, that's step 2.

Reply:
Absolutely, I wasn’t trying to suggest that this needs fixing before merging. Just something that we might want to cleanup at some point.